protocol/forge-scripts-deploy-in-go.md

```diff
@@ -0,0 +1,222 @@
+[OPStackManager design doc 60](https://github.com/ethereum-optimism/design-docs/pull/60) the importance
+of Modular configs and incremental deploy steps is outlined.
+
+To utilize those modular configs and deploy steps, we need a way to either 
+run and cache the individual Forge script functions (as outlined in 52),
+or integrate it closer such that no caching complexity is required.
+
+This document proposes the latter: run the Forge scripts within Go,
+such that we do not need multiple steps of Forge script sub-process to build a state for testing / genesis tooling.
+
+## Running Forge scripts in Go
+
+Forge scripts are really just solidity smart-contracts, with the addition of Forge cheat-codes.
+Running a contract in Go is relatively trivial: e.g. Cannon tests run the Cannon contracts in an instrumented Go EVM.
+
+To support the Forge cheatcodes, we need to emulate the cheatcode behavior, as the Geth EVM does not natively have support for it.
+Cheatcodes are essentially responses to `STATICCALL`s to a special system address, acting like a precompile,
+with functions that interact with the EVM environment.
+
+We do not have to support all Forge cheatcodes; just the subset used by the OP-Stack scripts would be sufficient.
+
+The most important cheat-codes are:
+- `vm.chainId`: change chain ID
+- `vm.load`, `vm.store`: storage getter/setter
+- `vm.etch`: write contract code
+- `vm.deal`: set balance
+- `vm.prank`: change sender
+- `vm.getNonce`/`vm.setNonce`: account nonce getter/setter
+- `vm.broadcast`,`vm.startBroadcast`, `vm.stopBroadcast`: capture calls as transaction candidates. (Can be no-op, if we do not want to go through Go for production deployments)
+- `vm.getCode`/`vm.getDeployedCode`: artifact inspection
+- `vm.env{Bool/Uint/Int/etc.}`: env var getters
+- `vm.keyExists{...}/parse{...}/writeJson/etc.`: encoding utils
+- `vm.addr`: priv key to addr
+- `vm.label`: name an address
+- `vm.dumpState`: export EVM state
+- `vm.loadAllocs`: import EVM state
+
+Note that with many of these, Go instrumentation can really improve integration:
+- `dumpState`/`loadAllocs`: no need to encode/decode state or read/write to disk -> faster Go tests
+- `getCode`/`getDeployedCode`: attach directly to artifacts, and *track which artifacts were used for a deployment*
+- `broadcast`: we can extend the deploy-tool functionality with transactions. Out of scope for now, but can be very useful.
+
+## Forge-artifacts as Go FS
+
+A Go FS is a simple filesystem abstraction: it provides read-only access to some source of files.
+A local directly can be wrapped into such FS, but also tarballs, or even data embedded in the Go binary, can be represented as Go FS.
+By using this FS abstraction, we can make the access to artifacts very simple, and "mount" the relevant FS into it.
+
+For Go tests, this would be the local FS.
+
+For Go genesis tooling, this might be a bundled FS, or one from a versioned release tarball.
```

Comment:
Are you proposing that this is used outside of testing, ie for L2 genesis generation?

Reply:
I see in the follow up section this is mentioned